File: java-ai-service/src/main/java/com/nontracey/aiservice/infra/SemanticCache.java
package com.nontracey.aiservice.infra;

import com.nontracey.aiservice.dto.Dtos.AskData;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;

/**
 * 语义缓存:question 向量与历史问题向量的 cosine 相似度超过阈值即视为命中,直接复用历史答案,
 * 省一次 LLM 调用。与 B/D 同策略(阈值 0.95),内存实现(dev);生产可换 Redis + 向量近邻。
 *
 * <p><b>架构位置</b>:infra 层。RagController 在 /api/ask 中先对问题向量化,查本缓存,
 * 命中则直接返回(并在 usage 标注 cache_hit=true),未命中才走检索 + 生成,最后回写缓存。
 *
 * <p><b>相似度判定</b>:阈值 0.95 很严格,只有"几乎同一问题的不同措辞"才会命中,
 * 避免把语义相近但意图不同的问题误判为同一问(宁可不命中,不可答非所问)。
 *
 * <p><b>实现取舍</b>:线性扫描 + synchronized,实现简单、正确性优先;条目少(dev)时性能足够,
 * 规模上来后应替换为 ANN 近邻索引。
 *
 * <p><b>同构映射</b>:对应 B 项目 {@code app/infra/cache.py}。
 */
@Component
public class SemanticCache {

    /** 命中判定阈值:cosine >= 0.95 视为同一问题。 */
    private static final double THRESHOLD = 0.95;
    /** 历史(问题向量, 答案)条目列表;synchronized 方法保证并发读写安全。 */
    private final List<Entry> entries = new ArrayList<>();

    /**
     * 查询缓存:线性遍历,返回第一条相似度达标的历史答案。
     *
     * @param qEmb 当前问题的 embedding
     * @return 命中时返回历史 AskData;未命中返回 null
     */
    public synchronized AskData get(float[] qEmb) {
        for (Entry e : entries) {
            if (cosine(qEmb, e.emb) >= THRESHOLD) return e.data;
        }
        return null;
    }

    /**
     * 写入缓存:记录(问题向量, 答案)。
     *
     * @param qEmb 问题 embedding
     * @param data 对应的 RAG 答案
     */
    public synchronized void put(float[] qEmb, AskData data) {
        entries.add(new Entry(qEmb, data));
    }

    /** 计算两个向量的余弦相似度;维度不一致或零向量时返回 0。 */
    private static double cosine(float[] a, float[] b) {
        if (a.length != b.length) return 0;
        double dot = 0, na = 0, nb = 0;
        for (int i = 0; i < a.length; i++) { dot += a[i] * b[i]; na += a[i] * a[i]; nb += b[i] * b[i]; }
        return (na == 0 || nb == 0) ? 0 : dot / (Math.sqrt(na) * Math.sqrt(nb));
    }

    /** 缓存条目:问题 embedding + 对应答案。 */
    private record Entry(float[] emb, AskData data) {}
}

File: dotnet-ai-service/Services/SemanticCache.cs
namespace DotnetAiService.Services;

/// <summary>语义缓存:question 向量与历史向量 cosine 相似度 &gt; 阈值即命中,省一次 LLM 调用。
/// 与 B 同策略(阈值 0.95),内存实现(dev);生产可换 Redis + 向量近邻。
/// <para>架构位置:挂在 /api/ask 端点的"检索之前"(见 Program.cs):
/// 护栏通过后先把问题向量化,命中缓存直接返回,跳过检索 + 生成整条链路。
/// 对应 B 项目 app/infra/cache.py、C 项目 infra/SemanticCache.java。</para>
/// <para>关键设计点 —— 命中判定:不是精确匹配,而是"语义近似":
/// 问题 embedding 与历史问题 embedding 逐一算余弦相似度,≥0.95 即视为同一问题。
/// 阈值权衡:0.95 足够拦下"换个说法问同一件事"(省 token),又不至于把相似但不同的问题混答;
/// 过高(如 0.99)命中率低,过低(如 0.9)会答非所问。</para>
/// <para>使用约束:只对非流式请求生效 —— SSE 流式回答无法缓存复用(Program.cs 中判断);
/// 缓存的是完整响应 payload(answer+sources+usage),命中时原样返回。</para>
/// <para>实现取舍:线性扫描 List,命中判定 O(n) —— 演示规模够用;
/// lock 保护并发读写(singleton 注册,多请求共享)。</para>
/// </summary>
public class SemanticCache
{
    /// <summary>命中阈值:余弦相似度 ≥0.95 判定为同一问题(与 B/C 保持一致)。</summary>
    private const double Threshold = 0.95;
    /// <summary>缓存条目:(问题向量, 完整响应 payload)。只增不删(演示用,无 TTL/淘汰策略)。</summary>
    private readonly List<(float[] Emb, object Payload)> _entries = new();
    /// <summary>并发锁:Get/Put 都在锁内,避免读到写一半的列表。</summary>
    private readonly object _lock = new();

    /// <summary>查缓存:线性扫描所有历史问题向量,首个相似度 ≥阈值 的条目即命中。</summary>
    /// <param name="qEmb">当前问题的 embedding(端点已算好,避免重复调 Embedding)。</param>
    /// <returns>命中返回缓存的 payload;未命中返回 null(调用方继续走检索+生成)。</returns>
    public object? Get(float[] qEmb)
    {
        lock (_lock)
        {
            foreach (var e in _entries)
                if (Cosine(qEmb, e.Emb) >= Threshold)
                    return e.Payload;
            return null;
        }
    }

    /// <summary>写缓存:生成成功后由端点调用,把问题向量与完整响应一起存入。</summary>
    public void Put(float[] qEmb, object payload)
    {
        lock (_lock) { _entries.Add((qEmb, payload)); }
    }

    /// <summary>余弦相似度(与 RagService.Cosine 同式):点积 / 模长乘积。
    /// 维度不一致或零向量返回 0,防御性处理避免异常。</summary>
    private static double Cosine(float[] a, float[] b)
    {
        if (a.Length != b.Length) return 0;
        double dot = 0, na = 0, nb = 0;
        for (int i = 0; i < a.Length; i++) { dot += a[i] * b[i]; na += a[i] * a[i]; nb += b[i] * b[i]; }
        return (na == 0 || nb == 0) ? 0 : dot / (Math.Sqrt(na) * Math.Sqrt(nb));
    }
}
